src/tsolvers/axdiffsolver/ArraySolver.cc

```diff
@@ -380,17 +380,20 @@ TRes ArraySolver::checkExtensionality() {
     for (auto const & entry : equalitiesToPropagate) {
         PTRef extensionalityClause = computeExtensionalityClause(entry.first, entry.second);
         assert(logic.isOr(extensionalityClause));
-        // TODO: handle conflicts better
-        bool allFalsified = true;
-        for (PTRef lit : logic.getPterm(extensionalityClause)) {
-            PTRef atom = logic.isNot(lit) ? logic.getPterm(lit)[0] : lit;
-            assert(logic.isEquality(atom));
-            allFalsified = allFalsified and (logic.isNot(lit) ? isSatisfied(atom) : isFalsified(atom));
-        }
+        // TODO: handle conflicts better: We pass literals twice now.
+        // TODO: looks like the above graph traversal could be terminated early if a clause is all-falsified
+        Pterm & literals = logic.getPterm(extensionalityClause);
```

Comment:
Not sure I understand the TODOs.

Reply:
> `TODO: handle conflicts better: We pass literals twice now.`

What I mean is that the literals of `extensionalityClause` are traversed twice, first in the `std::all_of` and then in the `for` loop on lines 396 -- 402.  Instead it could be an option to construct `explanation` speculatively and in case it is not needed, simply ignore it.

> `TODO: looks like the above graph traversal could be terminated early if a clause is all-falsified`

Elements are inserted into `equalitiesToPropagate` on line 372.  It looks like we could at that point already run the code on lines 381 -- 407, and if `allFalsified` is true, there's no need to continue the graph traversal, we could exit early and return the conflict.

Reply:
OK, got it.
I think it is better to keep them separate, because the case when all are falsified is the rare case. So most of the time, the speculation would not pay off and would be unnecessary overhead.

Reply:
Ok, but that only applies to the first TODO, right?

Reply:
I think I see what you mean now.
My last note is about the first TODO, and you are right that instead of collecting all `equalitiesToPropagate` first and checking them later, we could process them right away. No need to create the collection first.

Let's keep the TODO and revisit this in the next iteration/PR.